runtime-sdk/src/dispatcher.rs

```diff
@@ -525,14 +605,68 @@ impl<R: Runtime + Send + Sync> transaction::dispatcher::Dispatcher for Dispatche
         &self,
         rt_ctx: transaction::Context<'_>,
         batch: &mut TxnBatch,
-        _in_msgs: &[roothash::IncomingMessage],
+        in_msgs: &[roothash::IncomingMessage],
     ) -> Result<ExecuteBatchResult, RuntimeError> {
         let cfg = R::SCHEDULE_CONTROL.unwrap(); // Must succeed otherwise we wouldn't be here.
         let mut tx_reject_hashes = Vec::new();
 
         let mut result = self.execute_batch_common(
             rt_ctx,
             |ctx| -> Result<Vec<ExecuteTxResult>, RuntimeError> {
+                // Execute incoming messages.
+                let in_msgs_gas_limit = R::Core::remaining_in_msgs_gas(ctx);
+                let mut in_msgs_processed = 0usize;
+                for in_msg in in_msgs {
+                    let data = Self::decode_in_msg(in_msg).unwrap_or_else(|err| {
+                        warn!(ctx.get_logger("dispatcher"), "incoming message data malformed"; "id" => in_msg.id, "err" => ?err);
+                        IncomingMessageData::noop()
+                    });
+                    let tx = match data.ut.as_ref() {
```

Comment:
it should be synced up, other than the logic to determine how many messages fit within the incoming messages gas limit. unification runs into the same concerns as the transaction processing: during non-scheduling execution, we've thus far preferred to parse all txes first; during scheduling execution, we parse and execute one by one.

Reply:
oh although one thing about this: enforcing the incoming messages gas limit is not done under consensus. only the scheduler chooses

Reply:
> other than the logic to determine how many messages fit within the incoming messages gas limit

But the proposer does not communicate how many messages were chosen to executors. So if executors don't use the same logic for choosing messages, there will be discrepancies.

This would be different if we go with the design that emits incoming messages as transactions as then the proposer would decide how many to process (and executors would just need to verify that the messages are actually correct and in order as determined by the consensus layer).

Reply:
oh okay then